File: probs/usaco/mmind/mmind.cpp
// Solution by: Albert Ye
// AlphaStar 2020 Summer Camp: CC31AB
// Problem: mar10-mmind
#include <bits/stdc++.h>

using namespace std;

int main()
{
	int n;
	cin >> n;
	vector<pair<string, pair<int, int> > > v;
	for(int i = 0; i < n; i++)
	{
		string num;
		int c, w;
		cin >> num >> c >> w;
		v.push_back(make_pair(num, make_pair(c, w)));
	}
	// brute force for all numbers
	for(int num = 1000; num <= 9999; num++)
	{
		bool works = 1;
		// check if it matches the description for all hints
		for(int x = 0; x < n; x++)
		{
			// indices that have already been matched in answer and hint, respectively
			vector<bool> m1(n), m2(n);
			string tmp = to_string(num);
			int sm = 0, df = 0;
			for(int i = 0; i < 4; i++)
			{
				if(tmp[i] == v[x].first[i])
				{
					sm++;
					// found a match for both strings at i
					m1[i] = m2[i] = 1;
				}
			}
			for(int i = 0; i < 4; i++)
			{
				if(!m1[i])
				{
					for(int j = 0; j < 4; j++)
					{
						// check if answer is equal to hint at different indices
						if(!(m2[j]) && tmp[i] == v[x].first[j])
						{
							df++;
							m1[i] = m2[j] = 1;
						}
					}
				}
			}
			// check if the amount same/different positions is the same as described
			// if not, this number doesn't work
			if(sm != v[x].second.first || df != v[x].second.second)
			{
				works = 0;
				break;
			}
		}
		if(works)
		{
			// we've found an answer
			cout << num << endl;
			return 0;
		}
	}
	cout << "NONE" << endl;
	// O(900N) solution. ok
}

File: probs/usaco/mmind/mmind.py
"""
Solution by Albert Ye
AlphaStar 2020 Summer Camp: CC31AB
Problem: mar10-mmind
"""
n = int(input())
hint = [] # list of hints
for i in range(n) :
    # each hint has a number and the number of correct/wrong guesses
    num, c, w = input().split(' ')
    c = int(c)
    w = int(w)
    hint.append([num,c,w])
# simulate for all 4-digit numbers
for num in range(1000, 10000) :
    # easier to treat as string
    tmp = str(num)
    works = True
    for x in range(n) :
        # whether or not each digit of the potential answer has been found
        m1 = [False, False, False, False]
        # whether or not each digit of the hint has been found
        m2 = [False, False, False, False]
        # sum and difference
        sm = 0
        df = 0
        # in same position
        for i in range(4) :
            if tmp[i] == hint[x][0][i] :
                sm += 1
                m1[i] = m2[i] = True
        # in different position
        for i in range(4) :
            if not m1[i] :
                for j in range(4) :
                    if (not m2[j]) and tmp[i] == hint[x][0][j] :
                        df += 1
                        m1[i] = m2[j] = True
        # check if number in same/different position is as expected
        if sm != hint[x][1] or df != hint[x][2] :
            works = False
            break
    # if it's as expected, print it and end
    if works :
        print(num)
        exit()
# if there are no solutions print NONE
print('NONE')
